### ISSUE TITLE
bank_customers

### ISSUE THREAD

**Issue by yeshesa** [yeshesa/customer_classification.ipynb#1]
"""
bank_shap_explain.py
سكريبت كامل: preprocessing -> RandomForest (GridSearch) -> SHAP explanations (global + local)
مقدم باللغة العربية، مُعَدّ للتشغيل في بيئة محلية مثل VS Code.

تعديل: غيّر INPUT_FILE إلى اسم/مسار ملف الإكسل الخاص بك إذا لزم.
"""

import os
import sys
import warnings

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split, GridSearchCV, StratifiedKFold
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    roc_auc_score,
    f1_score,
    recall_score,
    precision_score,
    accuracy_score,
)
from imblearn.over_sampling import SMOTE

import shap

# تجاهل بعض التحذيرات لتجربة أنظف
warnings.simplefilter(action="ignore", category=FutureWarning)
warnings.simplefilter(action="ignore", category=UserWarning)

# ------------------------- إعداد المتغيرات -------------------------
INPUT_FILE = "bank_customers.xlsx"  # غيّر المسار إذا لزم
SHEET_NAME = 0  # رقم الورقة أو اسمها
RANDOM_STATE = 42
TEST_SIZE = 0.2

# اسماء أعمدة هدف محتملة — سيبحث السكربت عن أحد هذه الأسماء تلقائياً
TARGET_CANDIDATES = ["Churn", "churn", "Attrition", "Target", "ClosedAccount", "label"]

# عميل للتفسير المحلي — إما index من مجموعة الاختبار (int) أو 'first' لاختيار أول صف
LOCAL_CLIENT_INDEX = "first"  # غيّر إلى رقم صحيح مثل 5 لاختيار عميل محدد من X_test بعد التقسيم

# ------------------------- دوال مساعدة -------------------------
def find_target_column(df, candidates=TARGET_CANDIDATES):
    for c in candidates:
        if c in df.columns:
            return c
    return None

def get_feature_names_from_column_transformer(ct, numeric_cols, categorical_cols):
    """
    يستخرج أسماء الميزات بعد تطبيق ColumnTransformer (sklearn >= 1.0 يدعم get_feature_names_out).
    """
    try:
        # available in sklearn >= 1.0
        names = ct.get_feature_names_out()
        # get_feature_names_out تعيد أسماء بصيغة 'num__col' أو 'cat__onehot__col'
        # لكن تُستخدم مباشرة مع مصفوفة numpy المحولة
        return list(names)
    except Exception:
        # fallback: بناء يدوي
        feature_names = []
        # للأجزاء الرقمية
        feature_names.extend(numeric_cols)
        # للأجزاء الفئوية، استنبط أسماء one-hot
        ohe = ct.named_transformers_.get("cat")
        if ohe is not None and hasattr(ohe.named_steps.get("onehot"), "get_feature_names_out"):
            cats = categorical_cols
            onehot = ohe.named_steps["onehot"]
            ohe_names = list(onehot.get_feature_names_out(categorical_cols))
            feature_names.extend(ohe_names)
        else:
            # إن فشل كل شيء، أعد أسماء عامة
            feature_names.extend([f"cat_{i}" for i in range(len(categorical_cols))])
        return feature_names

# ------------------------- الخطوة 1: القراءة والاكتشاف -------------------------
if not os.path.exists(INPUT_FILE):
    print(f"خطأ: لم أجد الملف '{INPUT_FILE}' في المجلد الحالي ({os.getcwd()}).")
    sys.exit(1)

print("قراءة ملف الإكسل...")
df = pd.read_excel(INPUT_FILE, sheet_name=SHEET_NAME)

print(f"أعاد الملف {df.shape[0]} سطر و {df.shape[1]} عمود.")

# تحديد عمود الهدف تلقائياً
target_col = find_target_column(df)
if target_col is None:
    print("لم أجد عمود هدف معروف تلقائياً. الرجاء التأكد من وجود عمود باسم 'Churn' أو 'Attrition' أو ما شابه.")
    sys.exit(1)

print(f"تم التعرف على عمود الهدف: '{target_col}'")

# تحويل قيم الهدف النصية إلى 0/1 إن لزم
if df[target_col].dtype == object:
    unique_vals = df[target_col].dropna().unique()
    # شائع: 'Yes'/'No' أو 'Y'/'N'
    mapping = {}
    if set([str(x).lower() for x in unique_vals]) <= {"yes", "no"}:
        mapping = {"Yes": 1, "No": 0, "yes": 1, "no": 0}
    elif set([str(x).lower() for x in unique_vals]) <= {"y", "n"}:
        mapping = {"Y": 1, "N": 0, "y": 1, "n": 0}
    # إن لم نعرف، سنحاول تناسب القيم الثنائية إلى 0/1
    if mapping:
        df[target_col] = df[target_col].map(mapping)
    else:
        # حاول تحويل إلى عدد صحيح إذا أمكن
        try:
            df[target_col] = df[target_col].astype(int)
        except Exception:
            print("تحذير: العمود الهدف ليس ثنائيًا بمعرفة نصية. تأكد من القيم أو قم بتحويلها يدوياً.")
            # لا ننهِ التنفيذ؛ سنحاول الاستمرار

# فصل الميزات والهدف
X = df.drop(columns=[target_col])
y = df[target_col]

# التعرف على أنواع الأعمدة
numeric_cols = X.select_dtypes(include=["int64", "float64"]).columns.tolist()
categorical_cols = X.select_dtypes(include=["object", "category", "bool"]).columns.tolist()

print("الأعمدة الرقمية:", numeric_cols)
print("الأعمدة الفئوية:", categorical_cols)

# ------------------------- الخطوة 2: تقسيم البيانات -------------------------
print("تقسيم البيانات إلى تدريب/اختبار (stratify إذا أمكن)...")
if y.nunique() > 1:
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=TEST_SIZE, stratify=y, random_state=RANDOM_STATE
    )
else:
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=TEST_SIZE, random_state=RANDOM_STATE
    )

# ------------------------- الخطوة 3: تجهيز المعالج (Preprocessor) -------------------------
num_pipe = Pipeline(
    steps=[("imputer", SimpleImputer(strategy="median")), ("scaler", StandardScaler())]
)

cat_pipe = Pipeline(
    steps=[("imputer", SimpleImputer(strategy="most_frequent")), ("onehot", OneHotEncoder(handle_unknown="ignore"))]
)

preprocessor = ColumnTransformer(
    transformers=[("num", num_pipe, numeric_cols), ("cat", cat_pipe, categorical_cols)]
)

# fit preprocessor on training set to be able to get feature names later
print("تعلّم محول الميزات (fit preprocessor)...")
preprocessor.fit(X_train)

# أسماء الميزات بعد التحويل
feature_names = get_feature_names_from_column_transformer(preprocessor, numeric_cols, categorical_cols)
print(f"عدد الميزات بعد التحويل: {len(feature_names)}")

# تطبيق التحويل
X_train_t = preprocessor.transform(X_train)
X_test_t = preprocessor.transform(X_test)

# ------------------------- خطوة اختيار/موازنة الفئات (اختياري) -------------------------
# هنا نضع SMOTE كخيار (يمكن تعطيله بتعليق الأسطر التالية)
USE_SMOTE = False
if USE_SMOTE:
    print("تطبيق SMOTE على مجموعة التدريب...")
    sm = SMOTE(random_state=RANDOM_STATE)
    X_train_t, y_train = sm.fit_resample(X_train_t, y_train)

# ------------------------- الخطوة 4: تدريب نموذج RandomForest مع GridSearch -------------------------
print("بناء نموذج RandomForest وتحسين معاملاته باستخدام GridSearchCV...")
rf = RandomForestClassifier(random_state=RANDOM_STATE, class_weight="balanced")

param_grid = {
    "n_estimators": [100, 200],
    "max_depth": [6, 10, None],
    "min_samples_split": [2, 5],
}

cv = StratifiedKFold(n_splits=3, shuffle=True, random_state=RANDOM_STATE)
grid = GridSearchCV(rf, param_grid, scoring="roc_auc", cv=cv, n_jobs=-1, verbose=1)
grid.fit(X_train_t, y_train)

best_rf = grid.best_estimator_
print("أفضل معاملات:", grid.best_params_)

# ------------------------- الخطوة 5: تقييم النموذج -------------------------
print("تقييم النموذج على مجموعة الاختبار...")
y_pred = best_rf.predict(X_test_t)
y_proba = best_rf.predict_proba(X_test_t)[:, 1] if hasattr(best_rf, "predict_proba") else best_rf.predict(X_test_t)

metrics = {
    "AUC": roc_auc_score(y_test, y_proba),
    "F1": f1_score(y_test, y_pred),
    "Recall": recall_score(y_test, y_pred),
    "Precision": precision_score(y_test, y_pred),
    "Accuracy": accuracy_score(y_test, y_pred),
}
print("مقاييس الأداء:")
for k, v in metrics.items():
    print(f" - {k}: {v:.4f}")

# ------------------------- الخطوة 6: SHAP تفسير عام ومحلي -------------------------
print("حساب قيم SHAP باستخدام TreeExplainer...")
# shap يحتاج بيانات في الصيغة numpy أو DataFrame مع أسماء أعمدة
# نصنع DataFrame من X_test_t بأسماء feature_names إن أمكن
try:
    X_test_t_df = pd.DataFrame(X_test_t, columns=feature_names)
    X_train_t_df = pd.DataFrame(X_train_t, columns=feature_names)
except Exception:
    # إن تعذر، استخدم numpy arrays بدون أسماء
    X_test_t_df = pd.DataFrame(X_test_t)
    X_train_t_df = pd.DataFrame(X_train_t)

# TreeExplainer
explainer = shap.TreeExplainer(best_rf, data=X_train_t_df, feature_perturbation="tree_path_dependent")
# shap_values قد يكون مصفوفة (لـ binary returns list) أو مصفوفة ثنائية الأبعاد حسب نسخة shap/sklearn
shap_values = explainer.shap_values(X_test_t_df)

# التعامل مع الصيغة: نريد قيم class 1 (إيجابي/رحيل)
if isinstance(shap_values, list) or isinstance(shap_values, tuple):
    # shap_values[1] تمثل تأثير الفئة الموجبة عادةً
    shap_vals_pos = shap_values[1]
else:
    shap_vals_pos = shap_values

# --- رسم summary plot وحفظه كـ PNG ---
print("رسم SHAP summary plot وحفظه...")
plt.figure(figsize=(10, 6))
# استخدم shap.summary_plot ثم نحفظ الشكل الناتج
shap.summary_plot(shap_vals_pos, X_test_t_df, show=False)
plt.tight_layout()
plt.savefig("shap_summary.png", dpi=300)
plt.close()
print("تم حفظ shap_summary.png")

# --- حساب متوسط القيمة المطلقة لقيم SHAP لعمل bar chart ---
print("حساب أهمية الميزات (mean |SHAP|) ورسم شريطي...")
mean_abs_shap = np.abs(shap_vals_pos).mean(axis=0)
fi = pd.DataFrame({"feature": X_test_t_df.columns, "mean_abs_shap": mean_abs_shap})
fi = fi.sort_values("mean_abs_shap", ascending=False).reset_index(drop=True)

plt.figure(figsize=(10, 8))
sns.barplot(data=fi.head(30), x="mean_abs_shap", y="feature")
plt.title("Feature importance (mean |SHAP value|) — top 30")
plt.xlabel("mean |SHAP value|")
plt.tight_layout()
plt.savefig("shap_feature_importance_bar.png", dpi=300)
plt.close()
print("تم حفظ shap_feature_importance_bar.png")

# ------------------------- الخطوة 7: تفسير محلي (عميل محدد) -------------------------
# اختيار عميل محلي لشرح القرار
if LOCAL_CLIENT_INDEX == "first":
    local_idx = 0
elif isinstance(LOCAL_CLIENT_INDEX, int):
    local_idx = LOCAL_CLIENT_INDEX
else:
    try:
        local_idx = int(LOCAL_CLIENT_INDEX)
    except Exception:
        local_idx = 0

# تأكد من أن الفهرس صالح
if local_idx < 0 or local_idx >= X_test_t_df.shape[0]:
    print("مؤشر العميل المحلي خارج النطاق. سيتم اختيار العميل الأول بدلًا عنه.")
    local_idx = 0

print(f"إعداد تفسير محلي للعميل في موضع الاختبار: index={local_idx}")

# force_plot معمول له حفظ كـ HTML
shap.initjs()
force_plot = shap.force_plot(
    explainer.expected_value[1] if hasattr(explainer, "expected_value") and isinstance(explainer.expected_value, (list, np.ndarray)) else explainer.expected_value,
    shap_vals_pos[local_idx, :],
    X_test_t_df.iloc[local_idx, :],
    matplotlib=False,
)

html_filename = f"force_plot_client_{local_idx}.html"
shap.save_html(html_filename, force_plot)
print(f"تم حفظ التفسير المحلي كملف HTML: {html_filename}")

# ------------------------- خاتمة -------------------------
print("\nانتهى التنفيذ بنجاح.")
print("الملفات الناتجة في المجلد الحالي:")
for fname in ["shap_summary.png", "shap_feature_importance_bar.png", html_filename]:
    print(" -", fname)

print("\nملاحظات سريعة:")
print(" - افتح ملف HTML للتفسير المحلي في متصفح الويب لعرض force plot التفاعلي.")
print(" - إن أردت تفسيرًا لعميل آخر، غيّر LOCAL_CLIENT_INDEX في أعلى الملف إلى رقم الصف المطلوب ضمن مجموعة الاختبار.")
